File: Merge_Sort/Merge_Sort.js
function mergeSort(arr){
	let temp = new Array()
	divide(arr,temp,0,arr.length-1)
}

function divide(arr,temp,leftStart,rightEnd){
	if (leftStart >= rightEnd) {
		return
	}

	let middle = Math.floor((leftStart+rightEnd)/2)
	divide(arr,temp,leftStart,middle)
	divide(arr,temp,middle+1,rightEnd)
	mergeHalves(arr,temp,leftStart,rightEnd)
}

function mergeHalves(arr,temp,leftStart,rightEnd){

	let leftEnd = Math.floor((leftStart+rightEnd)/2)
	let rightStart = leftEnd + 1
	let size = rightEnd - leftStart + 1

	let index = leftStart
	let left = leftStart
	let right = rightStart

	while(left<=leftEnd && right<=rightEnd){
		if (arr[left] <= arr[right]) {
			temp[index] = arr[left]
			left++
		}else{
			temp[index] = arr[right]
			right++
		}

		index++
	}

	copyArray(arr,left,temp,index,leftEnd)
	copyArray(arr,right,temp,index,rightEnd)
	copyArray(temp,leftStart,arr,leftStart,rightEnd)

}

function copyArray(a,aBegin,b,bBegin,size){

	for (let i = aBegin,j=bBegin; i <=size; i++,j++) {
		b[j] = a[i]
	}
}

let arr = [10,9,8,7,6,5,4,3,2,1]
mergeSort(arr)
console.log(arr)

File: Merge_Sort/Merge_Sort.R
n = as.integer(readline())
arr = c()

for (i in 1:n) {
	arr[i] = as.integer(readline())
}

mergeSort<-function(arr){
	temp = rep(0,length(arr))
	arr = divide(arr,temp,1,length(arr))
	return(arr)
}

divide<-function(arr,temp,leftStart,rightEnd){

	if(leftStart >= rightEnd)
		return(arr);

	middle = floor((leftStart + rightEnd)/2)
	arr = divide(arr,temp,leftStart,middle)
	arr = divide(arr,temp,middle+1,rightEnd)
	mergeHalves(arr,temp,leftStart,rightEnd)
}

mergeHalves<-function(arr,temp,leftStart,rightEnd){

	leftEnd = floor((leftStart + rightEnd)/2)
	rightStart = leftEnd + 1
	size = rightEnd - leftStart + 1

	left = leftStart
	right = rightStart
	index = leftStart

	while(left<=leftEnd && right<=rightEnd){
		if(arr[left] <= arr[right]){
			temp[index] = arr[left]
			left = left + 1
		}else{
			temp[index] = arr[right]
			right = right + 1
		}
		index = index + 1

	}
	if(left<=leftEnd){
    temp = copyArray(arr,left,temp,index,leftEnd)
	}else{
    temp = copyArray(arr,right,temp,index,rightEnd)
	}
    arr = copyArray(temp,leftStart,arr,leftStart,rightEnd)

}

copyArray<-function(a,aBegin,b,bBegin,size){
  for (i in aBegin:size) {
    b[bBegin] = a[i]
    bBegin = bBegin + 1
  }
  return(b)
}

printArray<-function(arr){

	output = "["
	for (i in 1:length(arr)) {
		output = paste(output," ",arr[i]," ")
		if(i<length(arr))
			output = paste(output,",")
	}
	output = paste(output,"]")
	print(output)

}

arr = mergeSort(arr)
printArray(arr)
